staging/src/k8s.io/client-go/plugin/pkg/client/auth/azure/README.md

```diff
@@ -0,0 +1,68 @@
+# Azure Active Directory plugin for client authentication
+
+This plugin provides an integration with [Azure CLI](https://github.com/Azure/azure-cli) and Azure Active Directory device flow.
+
+## Usage
+
+### Microsoft Azure CLI 2.0
+
+1. Configure the `apiserver` to use the Azure Active Directory as an OIDC provider with following options
+
+   ```
+   --oidc-client-id="https://management.core.windows.net/" \
+   --oidc-issuer-url="https://sts.windows.net/TENANT_ID/"
+   --oidc-username-claim="sub"
+   ```
+
+   * Replace `TENANT_ID` with your tenant ID.
+
+2. Login with Azure CLI
+
+  ```
+  az login
+  ```
+
+3. Configure the `kubectl` to use the `azure` authentication provider
+
+   ```
+   kubectl config set-credentials "USER_NAME" --auth-provider=azure
+   ```
+
+   * Replace `USER_NAME` with your user name.
+
+### Device Flow
+
+1. Create an Azure native application following these [instructions](https://docs.microsoft.com/en-us/azure/active-directory/active-directory-app-registration)
+
+  Assign permissions to this application to access the `https://management.core.windows.net/"` audience.
+
+2. Configure the `apiserver` to use the Azure Active Directory as an OIDC provider with following options
+
+   ```
+   --oidc-client-id="https://management.core.windows.net/" \
+   --oidc-issuer-url="https://sts.windows.net/TENANT_ID/"
+   --oidc-username-claim="sub"
+   ```
+
+   * Replace `TENANT_ID` with your tenant ID.
+
+3. Configure the `kubectl` to use the `azure` authentication provider with using the registered application
+
+   ```
+   kubectl config set-credentials "USER_NAME" --auth-provider=azure \
+     --auth-provider-arg=client-id=APPLICATION_ID \
+     --auth-provider-arg=tenant-id=TENANT_ID \
+     --auth-provider-arg=audiance=https://management.core.windows.net/
+   ```
+
+   * Replace `USER_NAME`, `APPLICATION_ID` and `TENANT_ID` with the values of the registered application.
+
+ 4. The access token is acquired when first `kubectl` command is executed
+
+   ```
+   kubeclt get pods
+
+   To sign in, use a web browser to open the page https://aka.ms/devicelogin and enter the code DEC7D48GA to authenticate.
```

Comment:
There are two scenarios: 
1. The user logs in with `az login` which stores the tokens in `~/.azure/accessToken.json` file. The azure plugin will read the tokens from this file and refresh them if needed using the client_id also from the same file. 

2. If the plugin does not find any `az` file available. It will start the device flow by prompting to the use the device code retrieved from AAD and waiting for use to login. The user will login with AAD in a browser by providing the device code. Finally the plugin will receive the tokens directly from AAD.  

Note that in both cases the tokens are stored in the kubernetes configuration. Next time when the kubectl is executed, it will read the token from kubernetes config first.